src/customer/strategies/bolt/bolt-customer-strategy.ts

```diff
@@ -0,0 +1,119 @@
+import { MissingDataError, MissingDataErrorType, NotInitializedError, NotInitializedErrorType } from '../../../common/error/errors';
+import { BoltCheckout, BoltScriptLoader } from '../../../payment/strategies/bolt';
+import CustomerAccountRequestBody from '../../customer-account';
+import CustomerActionCreator from '../../customer-action-creator';
+import CustomerCredentials from '../../customer-credentials';
+import { CustomerRequestOptions } from '../../customer-request-options';
+import { GuestCredentials } from '../../guest-credentials';
+import CustomerStrategy from '../customer-strategy';
+
+export default class BoltCustomerStrategy implements CustomerStrategy {
+    private _boltClient?: BoltCheckout;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _boltScriptLoader: BoltScriptLoader,
+        private _billingAddressActionCreator: BillingAddressActionCreator,
+        private _customerActionCreator: CustomerActionCreator
+    ) {}
+
+    initialize(options: CustomerRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId } = options;
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const state = this._store.getState();
+        const paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
+
+        if (!paymentMethod || !paymentMethod.initializationData.publishableKey) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const { developerConfig, publishableKey } = paymentMethod.initializationData;
+
+        return this._boltScriptLoader.load(publishableKey, paymentMethod.config.testMode, developerConfig)
+            .then(boltCheckout => this._boltClient = boltCheckout)
+            .then(() => this._store.getState());
+    }
+
+    deinitialize(): Promise<InternalCheckoutSelectors> {
+        this._boltClient = undefined;
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    continueAsGuest(credentials: GuestCredentials, options?: CustomerRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId } = options || {};
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this.openCustomBoltCheckoutWidget(credentials.email, methodId)
+            .then(() => this._store.dispatch(this._billingAddressActionCreator.continueAsGuest(credentials, options)))
+            .then(() => this._store.getState());
+    }
+
+    signIn(credentials: CustomerCredentials, options?: CustomerRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId } = options || {};
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._store.dispatch(this._customerActionCreator.signInCustomer(credentials, options))
```

Comment:
Doesn't this mean that if a shopper chooses to sign into their BC account, and then refresh the page before they get a chance to complete whatever that's needed in the Bolt widget, they won't be prompted to complete the Bolt widget again (because the shopper is already signed in so they will be prompted to complete the next/shipping step instead)?

Reply:
The logged in shopper will have Bolt checkout modal open each time when open Checkout page and if the shopper has Bolt Account

Reply:
I think we have to be careful on this because you're assuming that the shopper wants to use Bolt. It might be that they intentionally close the Bolt modal. But they are prompted again everytime they reopen checkout or go back to the customer step.

Reply:
But we don't know why the user closed Bolt Checkout modal, maybe shopper want to get more stuff to make the order or so. We will open the modal only for those customer that has their own Bolt Account. Otherwise the modal will not be showed.

We discussed this case a couple times... And we made the decision to make it like this.

Reply:
I understand your team has made a decision. And I'm sorry if you feel like we're revisiting that decision. But considering that decision was made without consulting with our team, I think it's reasonable for us to question it and perhaps make adjustments to it if we both agree they are appropriate.

Anyway the reason why I brought this up is because you're making an assumption that shoppers with an existing Bolt account really want to use Bolt. Having a Bolt account does not mean the shopper wants to use it. It just means they have an account which they may choose to use it. Anyway, I guess you can say the shopper can just close it again if they really don't want use Bolt. That's fair enough. But what if the store has PayPal enabled and the shopper (with an existing Bolt account) clicks on the PayPal button on the cart page, authorises the payment on PayPal side and then get redirected to checkout, would they see the Bolt modal open automatically? If that's the case, I think we have a problem. At least, we have to make sure that if there's a payment in progress, don't try to encourage the shopper to abandon it and switch to use Bolt.

Reply:
Thanks @davidchin!

It's ok. We will discuss the issue and I or someone from my team will tell you about what we will doing next.

But it's a question for another task and the case has not to be a blocker for current task.

Reply:
> I understand your team has made a decision. And I'm sorry if you feel like we're revisiting that decision. But considering that decision was made without consulting with our team, I think it's reasonable for us to question it and perhaps make adjustments to it if we both agree they are appropriate.

@davidchin this is what these 2 PRs about, as long as the flow is a bit tricky, we made some decisions and started conversation with these 2 POCs

>I think we have to be careful on this because you're assuming that the shopper wants to use Bolt. It might be that they intentionally close the Bolt modal. But they are prompted again everytime they reopen checkout or go back to the customer step.

To my knowdledge, we have decided to introduce some flag in checkout session (the research on this feature is in the progress), that will be set whenever a buyer mananges popup, so in the scope of single quote they will not be bombarded with the popup over and overagain